src/main/java/usecases/join_public_lobby/JplInteractor.java

```diff
@@ -0,0 +1,127 @@
+package usecases.join_public_lobby;
+
+import entities.LobbyManager;
+import entities.Player;
+import entities.PlayerPoolListener;
+import entities.games.Game;
+import exceptions.EntityException;
+import usecases.GameDTO;
+import usecases.Response;
+
+/**
+ * Core class of the Join Public Lobby use case
+ * Given a player who would like to join a public lobby, adds this player
+ * to the matchmaking waiting pool in the entities. The thread blocks until
+ * the entity responds that the player was sorted into a game or that another
+ * use case was triggered to cancel this player's waiting. In both cases, the thread
+ * finishes execution and notifies the view model of the corresponding development
+ */
+public class JplInteractor implements JplInputBoundary {
+
+    private final LobbyManager lobbyManager;
+    private final JplOutputBoundary presenter;
+
+    /**
+     * Thread which executes the core logic of this use case
+     */
+    public class JplThread implements Runnable, PlayerPoolListener {
+
+        private Game game;
+        private boolean hasCancelled;
+        private final JplInputData data;
+
+        /**
+         *
+         * @param data Data passed into this use case
+         */
+        public JplThread (JplInputData data) {
+            this.data = data;
+            hasCancelled = false;
+        }
+
+        /**
+         * Called from another thread, sets game attribute. The run method
+         * reaches a point where is waits for either hasCancelled or game to be set
+         * @param game The game that the player has been sorted into
+         */
+        @Override
+        public void onJoinGamePlayer(Game game) {
+            this.game = game;
+        }
+
+        /**
+         * Called from another thread, sets hasCancelled to true. The run method
+         * reaches a point where is waits for either hasCancelled or game to be set
+         */
+        @Override
+        public void onCancelPlayer() {
+            hasCancelled = true;
+        }
+
+        /**
+         * Core logic of the use case
+         */
+        @Override
+        public void run() {
+            try {
+                // Throws IdInUseException, code after runs if this line succeeded
+                Player player = lobbyManager.createNewPlayer(data.getDisplayName(), data.getId());
+
+                // Add player to matchmaking pool and subscribe to hear updates
+                lobbyManager.addPlayerToPool(player, this);
+
+                // Notifies presenter that player was successfully added to pool
+                presenter.inPool(new JplOutputDataResponse(
+                        Response.getSuccessful("Player with ID " + player.getPlayerId() + " added to pool."),
+                        player.getPlayerId()
+                ));
+
+                // Block thread until an update is heard on the player in the pool
+                while (game == null && !hasCancelled) {
```

Comment:
Spin locks are usually inefficient, so we should try to remove them in the future. As for now, this approach shows the core idea which is what we need